Add docker collector for metrics provider

L3n41c opened the pull request:
### What does this PR do?

Implement the docker collector for metrics provider

### Motivation

### Additional Notes

Built on top of #9403 

### Describe how to test your changes

Run the generic `container` check on a Windows host with docker.

For Linux, it is a bit more complicated because the `system` collector always takes precedence. To disable it, run the Agent without mounting `/proc` to `/host/proc`, the `system` provider will be automatically disabled.

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

L3n41c committed:
Implement the docker collector for metrics provider

vboulineau commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,99 @@
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       dockerCollectorID,
+		priority: 1,
+		runtimes: allLinuxRuntimes,
```
This collector only support docker runtime.

L3n41c replied:
Now fixed.

vboulineau commented on pkg/util/containers/v2/metrics/collector_docker_linux.go:
```diff
@@ -0,0 +1,112 @@
+		Limit:      util.Float64Ptr(float64(memStats.Limit)),
+	}
+
+	if rss, found := memStats.Stats["rss"]; found {
```
IIRC, the `Stats` contain almost all cgroups memory data so you may be able to fill more fields?

L3n41c replied:
Here is the full content of the `Stats` map:
```go
map[string]uint64{
"active_anon":0x14a000,
"active_file":0x0,
"cache":0x0,
"dirty":0x0,
"hierarchical_memory_limit":0xaa12b000,
"hierarchical_memsw_limit":0x0,
"inactive_anon":0x0,
"inactive_file":0x0,
"mapped_file":0x0,
"pgfault":0x56a,
"pgmajfault":0x0,
"pgpgin":0x37b,
"pgpgout":0x20b,
"rss":0x142000,
"rss_huge":0x0,
"total_active_anon":0x14a000,
"total_active_file":0x0,
"total_cache":0x0,
"total_dirty":0x0,
"total_inactive_anon":0x0,
"total_inactive_file":0x0,
"total_mapped_file":0x0,
"total_pgfault":0x56a,
"total_pgmajfault":0x0,
"total_pgpgin":0x37b,
"total_pgpgout":0x20b,
"total_rss":0x142000,
"total_rss_huge":0x0,
"total_unevictable":0x0,
"total_writeback":0x0,
"unevictable":0x0,
"writeback":0x0}
```

Here are the fields of the `ContainerMemStats` structure:
```go
type ContainerMemStats struct {
        // Common fields
        UsageTotal   *float64
        KernelMemory *float64
        Limit        *float64
        Softlimit    *float64

        // Linux-only fields
        RSS       *float64
        Cache     *float64
        Swap      *float64
        OOMEvents *float64 // Number of events where memory allocation failed

        // Windows-only fields
        PrivateWorkingSet *float64
        CommitBytes       *float64
        CommitPeakBytes   *float64
}
```

The non-Windows fields that are not set yet are `KernelMemory`, `SoftLimit`, `Swap` and `OOMEvents`.
Do you think those fields could be computed from some of the `Stats` values ?

vboulineau replied:
You may need to handle both `cgroupv1` and `cgroupv2` because the fields are different and Docker reports them "raw":
https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1416
https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1510

I forgot that we did not put everything in the `ContainerMemStats` (from cgroups detailed stats), so indeed not a lot to get.

KernelMemory is mapped here for cgroups v2 (not available in v1): https://github.com/DataDog/datadog-agent/blob/7c519f08e2bb1bf20e409ba0e41e26663057eee0/pkg/util/cgroups/cgroupv2_memory.go#L40
So you could reuse the logic.

Swap is not returned by Docker regardless of cgroup versions.

OOMEvents (==FailCnt) is returned as a structured field by Docker here: https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L1486

From `docker run` documentation:
```
      --memory-reservation bytes       Memory soft limit
```

SoftLimit seems to correspond to the `MemoryReservation` field, but not entirely sure: https://github.com/moby/moby/blob/ba1629333018522ae09d55a564ec6a7b16abcbb8/daemon/daemon_unix.go#L89

L3n41c replied:
Thank you very much Vincent for all those details !

* I’ve added the computation of `KernelMemory` based on `kernel_stack` and `slab`.
* I’ve filled `OOMEvents` with `Failcnt`.

About the memory soft limit, I tried to start a container with a “magic” value for the `--memory-reservation`, but I couldn’t find that figure back from the docker stats API:

```
$ docker run --name sleep --memory-reservation 12345678 --memory 23456789 -d ubuntu:20.04 sleep infinity

c621d20d0e11ea52250e8ed0f65fa177ecaf53646b65400c3a61764ba127b3f4


$ curl -s --unix /var/run/docker.sock 'http://unix/containers/c621d20d0e11ea52250e8ed0f65fa177ecaf53646b65400c3a61764ba127b3f4/stats?stream=false' | jq .memory_stats

{
  "usage": 327680,
  "max_usage": 4558848,
  "stats": {
    "active_anon": 0,
    "active_file": 0,
    "cache": 0,
    "dirty": 0,
    "hierarchical_memory_limit": 23453696,
    "hierarchical_memsw_limit": 46911488,
    "inactive_anon": 0,
    "inactive_file": 0,
    "mapped_file": 0,
    "pgfault": 1287,
    "pgmajfault": 0,
    "pgpgin": 924,
    "pgpgout": 936,
    "rss": 0,
    "rss_huge": 0,
    "total_active_anon": 0,
    "total_active_file": 0,
    "total_cache": 0,
    "total_dirty": 0,
    "total_inactive_anon": 0,
    "total_inactive_file": 0,
    "total_mapped_file": 0,
    "total_pgfault": 1287,
    "total_pgmajfault": 0,
    "total_pgpgin": 924,
    "total_pgpgout": 936,
    "total_rss": 0,
    "total_rss_huge": 0,
    "total_unevictable": 0,
    "total_writeback": 0,
    "unevictable": 0,
    "writeback": 0
  },
  "limit": 23453696
}
```

I’ve started my container with:
* `--memory 23456789`
* `--memory-reservation 12345678`

whereas I could find the `--memory` value rounded up to the closest multiple of 4 Ki in `.memory_stats.limit` and `.memory_stats.stats.hierarchical_memory_limit`, I cannot find any field with a value close to the value of `--memory-reservation`.
Do you think this could also come from a cgroup v1 limitation ?

vboulineau replied:
No it's not a limitation it's just that Docker does not report it because it comes from another file `memory.soft_limit_in_bytes` (and `memory.low` cgroupv2).

But I think the value is the one from `MemoryReservation` from **Container Spec**. However I'm not sure pulling this metric is worth triggering another API Call.

vboulineau commented on pkg/util/containers/v2/metrics/collector_docker_linux.go:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	for _, blkioStatEntry := range ioStats.IoServiceBytesRecursive {
+		deviceName := fmt.Sprintf("%d:%d", blkioStatEntry.Major, blkioStatEntry.Minor)
```
We tend to avoid reporting `major:minor` as device name and report the device name using the mapping.
The idea would be to see if we can re-use the `getDiskDeviceMapping` code in `pkg/util/containers/v2/metrics/collector_disk_linux.go`.

L3n41c replied:
Makes sense !
`convertIOStats` is now leveraging `getDiskDeviceMapping`.

vboulineau commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func newDockerCollector() (*dockerCollector, error) {
+	du, err := docker.GetDockerUtil()
```
Thinking that it should check first if the Docker feature is activated or not

L3n41c replied:
Check added !

vboulineau commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,99 @@
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       dockerCollectorID,
+		priority: 1,
+		runtimes: []string{"RuntimeNameDocker"},
```
typo: extra quotes

L3n41c replied:
🤦🏻 Indeed !
Now fixed !

L3n41c committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into lenaic/docker_metrics_provider

L3n41c committed:
CR fix

ahmed-mez commented on pkg/util/containers/v2/metrics/collector_ecs_fargate.go:
```diff
@@ -110,13 +110,13 @@ func convertEcsStats(ecsStats *v2.ContainerStats) *ContainerStats {
 }
 
-func convertCPUStats(cpuStats *v2.CPUStats) *ContainerCPUStats {
+func convertEcsCPUStats(cpuStats *v2.CPUStats) *ContainerCPUStats {
```
Let's also rename it `convertDockerCPUStats` in the docker collector, wondering if we should have each collector in a separate package.

L3n41c replied:
Let me move each collector to a dedicated sub-package. That sounds cleaner.

ahmed-mez replied:
we can do it after https://github.com/DataDog/datadog-agent/pull/9749 is merged

davidor replied:
+1 to moving each collector to its dedicated sub-package. We can do it in a separate PR if you prefer.

vboulineau replied:
Note that collectors are not in sub-packages to avoid registration + circular dependencies issues

L3n41c replied:
PR updated with all collectors in their dedicated sub-package.
It required to change the visibility of some symbols and to create an additional package to handle registration without circular dependencies. That extra package isn’t super elegant. :-(

L3n41c committed:
Move each provider in its dedicated subpackage

davidor commented on pkg/util/containers/v2/metrics/collector_docker_windows.go:
```diff
@@ -0,0 +1,59 @@
+
+func convertCPUStats(cpuStats *types.CPUStats) *ContainerCPUStats {
+	return &ContainerCPUStats{
+		Total:  util.Float64Ptr(float64(100 * cpuStats.CPUUsage.TotalUsage)),
```
I think it'd be nice to add a comment that explains why `100 *` is needed (what are the units of the usage in Windows).

L3n41c replied:
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-ff1cf6879822b2f78e4d27ad3529ed3bd61f2a6dc12c5819baf1145a1a757f6fR32

davidor commented on pkg/util/containers/v2/metrics/collector_docker_windows.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/docker/docker/api/types"
+)
+
+func convertContainerStats(stats *types.Stats) *ContainerStats {
```
Because this collector also works on Windows, I think you need to add it here: https://github.com/DataDog/datadog-agent/blob/1b8a06d6fc5e59f2834f7408659e2eb28f9a182e/pkg/autodiscovery/listeners/environment.go#L78

L3n41c replied:
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/files#diff-fb673103ed71b2f0eb00bcdabb898950e408964c5ac68e4fb6af37eaa1d5ed0b

L3n41c committed:
CR fix

L3n41c committed:
Add a cache in front of docker stats API

L3n41c committed:
Use networks ifname map

L3n41c committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into lenaic/docker_metrics_provider

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/docker"
+	"github.com/docker/docker/api/types"
```
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-289ca0fe17b5c1d9e5e96120563c069d7f3890c04cae4f3774751e9072819986R16

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+	return dockerCollectorID
+}
+
+func (d *dockerCollector) GetContainerStats(containerID string, caccheValidity time.Duration) (*ContainerStats, error) {
```
Fixed: https://github.com/DataDog/datadog-agent/commit/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-289ca0fe17b5c1d9e5e96120563c069d7f3890c04cae4f3774751e9072819986R67

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+func init() {
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       dockerCollectorID,
+		priority: 1,
```
Fixed: https://github.com/DataDog/datadog-agent/commit/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-289ca0fe17b5c1d9e5e96120563c069d7f3890c04cae4f3774751e9072819986R32

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker_linux.go:
```diff
@@ -0,0 +1,135 @@
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/docker/docker/api/types"
```
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-eb87191ae6a1a0b5da8cc5cdffa9d9dd18e6a485ae2f80c9e94e5d79491878b4R15

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker_linux.go:
```diff
@@ -0,0 +1,135 @@
+
+func convertPIDStats(pidStats *types.PidsStats) *ContainerPIDStats {
+	return &ContainerPIDStats{
+		ThreadCount: util.Float64Ptr(float64(pidStats.Current)),
```
Good point !
I migrated to `UIntToFloatPtr` everywhere it was possible: https://github.com/DataDog/datadog-agent/pull/9612/commits/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-eb87191ae6a1a0b5da8cc5cdffa9d9dd18e6a485ae2f80c9e94e5d79491878b4R36-R40

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker_linux_test.go:
```diff
@@ -0,0 +1,226 @@
+	"os"
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
```
I think you’re right: it’s cleaner like that.
I usually let `goimports` manage the `import` statement for me.
I should investigate how to configure my IDE to pass the proper value of `-local` to `goimports` depending on the project.

Anyway… fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/32bc1af90ab330d45edb7591ef0458b24a1c0eed#diff-509107831c8b8208e425ca0fe37dd81dec69cde93602eb810094ab71d76a7d9eR16-R17

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+	return convertContainerStats(&stats.Stats), nil
+}
+
+func (d *dockerCollector) GetContainerNetworkStats(containerID string, cacheValidity time.Duration, networks map[string]string) (*ContainerNetworkStats, error) {
```
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/6f5625d957f731f1a12011ed7bdaf5517e836db0

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+	return dockerCollectorID
+}
+
+func (d *dockerCollector) GetContainerStats(containerID string, caccheValidity time.Duration) (*ContainerStats, error) {
```
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/commits/75a226af2956bbdd6cd1c82c3a3a88684829c115

L3n41c commented on pkg/util/containers/v2/metrics/collector_docker.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	du, err := docker.GetDockerUtil()
+	if err != nil {
```
Fixed: https://github.com/DataDog/datadog-agent/pull/9612/files#diff-289ca0fe17b5c1d9e5e96120563c069d7f3890c04cae4f3774751e9072819986R58

L3n41c committed:
Fix

vboulineau commented on pkg/collector/corechecks/containers/generic/check.go:
```diff
@@ -14,6 +14,9 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics"
+
+	// Register all the metrics collectors
+	_ "github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/collectors"
```
That's the main reason why the collectors were not in a dedicated package, I am not a big fan of this pattern, it caused multiple issues with the previous container provider design.

L3n41c replied:
I managed to get rid of that by restructuring a little bit the package structure and introducing [this package facade](https://github.com/DataDog/datadog-agent/pull/9612/files#diff-ac4d9a82ecab221f3444c208275d304d1c841bcd122555890cb71900ea7169f0).

L3n41c committed:
Split the `pkg/util/containers/v2/metrics/` package

and introduce a facade to remove the need of registering collector packages and break circular dependency

ahmed-mez commented on pkg/util/containers/v2/metrics/docker/collector_docker.go:
```diff
@@ -0,0 +1,140 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
nit: we can simplify the file names since we moved each collector in a sub-package, `mv collector_docker.go collector.go` (it applies to every collector)

L3n41c replied:
Done!

ahmed-mez commented on pkg/util/containers/v2/metrics/docker/collector_docker.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// stats returns stats by container ID, it uses an in-memory cache to reduce the number of api calls.
+// Cache expires every 2 minutes and can also be invalidated using the cacheValidity argument.
```
```suggestion
// Cache expires every 10 seconds and can also be invalidated using the cacheValidity argument.
```

L3n41c replied:
Fixed

ahmed-mez commented on pkg/util/containers/v2/metrics/docker/collector_docker_test.go:
```diff
@@ -0,0 +1,176 @@
+
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"
```
```suggestion
	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"

```

L3n41c replied:
Fixed

ahmed-mez commented on pkg/util/containers/v2/metrics/docker/collector_docker_windows.go:
```diff
@@ -0,0 +1,62 @@
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"
```
```suggestion
	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"

```

L3n41c replied:
Fixed

ahmed-mez commented on pkg/util/containers/v2/metrics/docker/collector_docker_windows_test.go:
```diff
@@ -0,0 +1,132 @@
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"
```
```suggestion
	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics/provider"

```

L3n41c replied:
Fixed
